### ISSUE TITLE
castxml: Illegal instruction: 4

### ISSUE THREAD

**Issue by iMichka** [Homebrew/legacy-homebrew#49829]
Hi

I have some trouble using the castxml bottles.

Here is what I get on some machines (and especially in Travis builds), by just running the binary:

```7814 Illegal instruction: 4```

I isolated to problem as being an architecture problem:

Works: OS X 10.11.3, Intel(R) Core(TM) i5-4278U CPU @ 2.60GHz
Fails: OS X (?, Travis), Intel(R) Xeon(R) CPU X5570 @ 2.93GHz
Fails: OS X 10.11.3, Intel(R) Core(TM)2 Duo CPU T9900 @ 3.06GHz

I also tried the different travis xcode images, and all have the same problem. On the Core2duo I own, rebuilding castxml from source produces a castxml which works.

I tried to do some autopsy of the failure with lldb on the Core2duo:

```
michkapopoff$ lldb castxml
(lldb) target create "castxml"
Current executable set to 'castxml' (x86_64).
(lldb) r
Process 82624 launched: '/usr/local/bin/castxml' (x86_64)
Process 82624 stopped
* thread #1: tid = 0xc59047, 0x000000010069e1fc castxml`_GLOBAL__sub_I_CommandLine.cpp + 232, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x000000010069e1fc castxml`_GLOBAL__sub_I_CommandLine.cpp + 232
castxml`_GLOBAL__sub_I_CommandLine.cpp:
 ->  0x10069e1fc <+232>: vxorps %xmm0, %xmm0, %xmm0
    0x10069e200 <+236>: vmovups %xmm0, 0x10e41b0(%rip)    ; HLOp + 16
    0x10069e208 <+244>: leaq   0xe2c35d(%rip), %rax      ; ""
    0x10069e20f <+251>: movq   %rax, 0x10e41c2(%rip)     ; HLOp + 48
(lldb)
```

I am not sure if this how bottles are built with homebrew. Maybe there are some special flags to pass down to the compiler? As rebuilding from source on Core2Duo allowed me to make it work, I am quite confident there is a way to build binaries that work on all the architectures.

Adding @bradking, @thewtex as they may have an idea of what is going on here.

**Comment by thewtex** [comment]
@iMichka do the binaries here:

  https://midas3.kitware.com/midas/item/318762

work on the same systems?

The extra flags used to build them are here:

  https://github.com/thewtex/CastXMLSuperbuild/blob/df2997d9baca048477fe4c11a35c818343bf36b8/CMakeLists.txt#L30-L35

with LLVM 3.7.0.

**Comment by iMichka** [comment]
Yes the binaries you built are working.

I tried using ```-DCMAKE_OSX_DEPLOYMENT_TARGET:STRING=10.9``, and got the following build warning:


```
CMake Warning at /usr/local/Cellar/cmake/3.4.3/share/cmake/Modules/Platform/Darwin-Initialize.cmake:97 (message):
  CMAKE_OSX_DEPLOYMENT_TARGET is '10.9' but the matching SDK does not exist
  at:
   "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk"

  Instead using SDK:

 "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk".
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.4.3/share/cmake/Modules/CMakeSystemSpecificInitialize.cmake:18 (include)
  CMakeLists.txt:17 (project)
```

The binary was built, but it is still crashing. I guess you had the right SDK installed. I'll need to grab that one somewhere to try it again. Not sure we can use an older SDK on homebrew though. I personally do not care about the Core2Duo, but more about the Xeons used on Travis.

**Comment by mikemcquaid** [comment]
> Maybe there are some special flags to pass down to the compiler? As rebuilding from source on Core2Duo allowed me to make it work, I am quite confident there is a way to build binaries that work on all the architectures.

Homebrew does pass the CFLAGS to build generic binaries (e.g. for Core2Duo) so there's something in the buildsystem that's not respecting that.

**Comment by apjanke** [comment]
This might actually be something in Homebrew, not just CastXML. Or at least, not simply the arch flags in the formula build.

TL;DR: Maybe the bottle build is statically linking some libs that were built for newer CPUs for some reason, maybe from non-default-option dependencies?

In superenv, from what I can tell, we're not using `CFLAGS` specifically; it's adding `-march=core2` to the `clang` or `clang++` command line arguments. I did a `brew --build-bottle castxml` and the flags showed up as expected. 

The bad instruction is `VXORPS`.

```
* thread #1: tid = 0xc59047, 0x000000010069e1fc castxml`_GLOBAL__sub_I_CommandLine.cpp + 232, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x000000010069e1fc castxml`_GLOBAL__sub_I_CommandLine.cpp + 232
castxml`_GLOBAL__sub_I_CommandLine.cpp:
 ->  0x10069e1fc <+232>: vxorps %xmm0, %xmm0, %xmm0
```

That [Xeon(R) CPU X5570](http://ark.intel.com/products/37111/Intel-Xeon-Processor-X5570-8M-Cache-2_93-GHz-6_40-GTs-Intel-QPI) looks like it's a Nehalem, and that [`VXORPS` instruction](http://www.felixcloutier.com/x86/XORPS.html)) is from [AVX](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions), which was added in Sandy Bridge. Perhaps because this is an instruction set extension, and not a default part of the instruction set, `clang` is testing for support for it separately, maybe using CPUID feature flags. Or something.

However, I couldn't get a no-`vxorps` build of `castxml` on my system, even with passing different `--bottle-arch` flags, or modifying Homebrew to explicitly pass `-mno-avx`. Those ought to work, it seems.

But here's a thing: castxml `depends_on "llvm" => "with-clang"`. That's a non-default option, which means `llvm` will be built from source as part of the bottle build, right?

In the build logs, I can see the `castxml` binary being statically linked with several library files from the brewed `llvm`.

```
superenv executed: clang++ -pipe -w -Os -mno-avx -fno-exceptions -fno-rtti -std=c++11 -DNDEBUG -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/castxml.dir/castxml.cxx.o CMakeFiles/castxml.dir/Detect.cxx.o CMakeFiles/castxml.dir/Output.cxx.o CMakeFiles/castxml.dir/RunClang.cxx.o CMakeFiles/castxml.dir/Utils.cxx.o -o ../bin/castxml ../lib/libcxsys.a -lclangFrontend -lclangDriver -lclangSerialization -lclangParse -lclangSema -lclangAnalysis -lclangEdit -lclangAST -lclangLex -lclangBasic /usr/local/opt/llvm/lib/libLLVMOption.a /usr/local/opt/llvm/lib/libLLVMBitReader.a /usr/local/opt/llvm/lib/libLLVMSupport.a /usr/local/opt/llvm/lib/libLLVMAArch64CodeGen.a /usr/local/opt/llvm/lib/libLLVMAArch64AsmPrinter.a /usr/local/opt/llvm/lib/libLLVMAArch64AsmParser.a /usr/local/opt/llvm/lib/libLLVMAArch64Desc.a /usr/local/opt/llvm/lib/libLLVMAArch64Info.a /usr/local/opt/llvm/lib/libLLVMAArch64Disassembler.a /usr/local/opt/llvm/lib/libLLVMARMCodeGen.a /usr/local/opt/llvm/lib/libLLVMARMAsmPrinter.a /usr/local/opt/llvm/lib/libLLVMARMAsmParser.a /usr/local/opt/llvm/lib/libLLVMARMDesc.a /usr/local/opt/llvm
```

And on my system, some of those `.a` files from `llvm` themselves contain `vxorps` instructions.

```
$ for i in *; do str=$(otool -tvV $i | grep vxorps | head -1); if [[ -n "$str" ]]; then echo "$i: $str";ls -l $i; fi; done
BugpointPasses.dylib: 0000000000001c67	vxorps	%xmm0, %xmm0, %xmm0
-r--r--r-- 1 janke admin 17504 Mar  8 17:56 BugpointPasses.dylib
LLVMHello.dylib: 0000000000001d1d	vxorps	%xmm0, %xmm0, %xmm0
-r--r--r-- 1 janke admin 16920 Mar  8 17:45 LLVMHello.dylib
error: /Library/Developer/CommandLineTools/usr/bin/otool: can't map file: clang (Invalid argument)
libLLVMAArch64AsmParser.a: 000000000000005d	vxorps	%xmm0, %xmm0, %xmm0
-r--r--r-- 1 janke admin 298880 Mar  8 17:56 libLLVMAArch64AsmParser.a
libLLVMAArch64AsmPrinter.a: 0000000000001af8	vxorps	%xmm0, %xmm0, %xmm0
-r--r--r-- 1 janke admin 337216 Mar  8 17:56 libLLVMAArch64AsmPrinter.a
libLLVMAArch64CodeGen.a: 0000000000001af5	vxorps	%xmm0, %xmm0, %xmm0
-r--r--r-- 1 janke admin 1596464 Mar  8 17:56 libLLVMAArch64CodeGen.a
libLLVMAArch64Desc.a: 00000000000000b9	vxorps	%xmm0, %xmm0, %xmm0
-r--r--r-- 1 janke admin 488144 Mar  8 17:56 libLLVMAArch64Desc.a
```

I think that is dependent on the architecture/cpu options that `llvm` is built with, and the `-m` options passed to `clang` at the point of building `castxml` won't matter. I found that when I did `brew rm llvm castxml` and then `brew install --build-bottle castxml`, the [`llvm` it built from source as a dependency](https://gist.github.com/07fe3f2de8b98ead198b) still had `-march=native`: `superenv executed: clang++ -pipe -w -Os -march=native...`

There are some other bottled formulae on my machine which also have `vxorps` instructions: `jerasure`, `gf-complete`, `gsl`, `v8`, `node`, `x264`. Perhaps they are also picking them up from dependencies with non-default options which got built with `-march=native`. Or perhaps there are some static libs in the OS X SDK or something else which were built against an AVX-capable processor?

```
[/usr/local/bin on ⇄ master ±]
$ for i in *; do str=$(otool -tvV $i | grep vxorps | head -1); if [[ -n "$str" ]]; then echo "$i: $str";ls -l $i; fi; done
error: /Library/Developer/CommandLineTools/usr/bin/otool: can't open file: 2to3-3.4 (No such file or directory)
castxml: 0000000100012318	vxorps	%xmm0, %xmm0, %xmm0
lrwxr-xr-x 1 janke admin 47 Mar  8 20:30 castxml -> ../Cellar/castxml/0.1+git20160202-1/bin/castxml
decoder: 0000000100002278	vxorps	%xmm1, %xmm1, %xmm1
lrwxr-xr-x 1 janke admin 34 Dec 29 17:49 decoder -> ../Cellar/jerasure/2.0/bin/decoder
encoder: 0000000100002b94	vxorps	%xmm1, %xmm1, %xmm1
lrwxr-xr-x 1 janke admin 34 Dec 29 17:49 encoder -> ../Cellar/jerasure/2.0/bin/encoder
ex: 00000001000081e5	vxorps	%xmm0, %xmm0, %xmm0
...
```

Could be something about whether `--build-bottle` is propagated to the dependencies that are built from source, or when the dependency check & installation is done.

I've kicked off a bottle build with a dummy PR #49892 so we can see logs from the Jenkins server where the bottles are built, in case that helps.

**Comment by apjanke** [comment]
That may be it. The counter-test worked. After doing a `brew install --build-bottle llvm --with-clang`, I got xvorps-free `*.a` files under `/usr/local/opt/llvm`, and then `brew install --build-bottle castxml` resulted in a xvorps-free `castxml` binary.

```
$ brew install --build-bottle castxml
==> Using the sandbox
==> Downloading https://mirrors.ocf.berkeley.edu/debian/pool/main/c/castxml/cast
Already downloaded: /Library/Caches/Homebrew/castxml-0.1+git20160202-1.tar.xz
==> cmake .. -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG
==> make install
🍺  /usr/local/Cellar/castxml/0.1+git20160202-1: 69 files, 33.5M, built in 13 seconds
[~]
$ otool -tvV /usr/local/bin/castxml | grep vxorps | head -1
[~]
$
```

I think this means we need to do something about propagating `--build-bottle` to dependencies which are built from source as part of a bottle installation. (Including from-source dependencies pulled in via intermediate bottles in the dependency tree, if such a thing is possible.) Should only apply to formulae which include statically-linked libraries, or other compile-time-determined CPU-dependent stuff, but we don't know which those are, so may need to just do it for all.

**Comment by apjanke** [comment]
The [bottle build](http://bot.brew.sh/job/Homebrew%20Pull%20Requests/42955/) finished, but the build log for dependencies including `llvm` was too [truncated in the Jenkins web page](http://bot.brew.sh/job/Homebrew%20Pull%20Requests/42955/version=el_capitan/testReport/brew-test-bot/el_capitan/install___only_dependencies_castxml/) to be useful.

```
Passed

brew-test-bot.el_capitan.install --only-dependencies castxml
Took 18 min.
add description
Standard Output

        truncated output to 1MB:
ch64Utils.a
[ 75%] Built target LLVMAArch64Utils
[ 75%] Building CXX object lib/Target/AArch64/MCTargetDesc/CMakeFiles/LLVMAArch64Desc.dir/AArch64TargetStreamer.cpp.o
cd /tmp/llvm20160308-5800-5wctd7/lib/Target/AArch64/MCTargetDesc && /usr/local/Library/ENV/4.3/clang++   -DGTEST_HAS_RTTI=0 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/tmp/llvm20160308-5800-5wctd7/lib/Target/AArch64/MCTargetDesc -I/tmp/llvm20160308-5800-ya3k7m/llvm-3.6.
...[truncated 1047617 chars]...
:
  echo /usr/local/opt/llvm/lib/python2.7/site-packages >> /usr/local/lib/python2.7/site-packages/llvm.pth
  mkdir -p /Users/brew/Jenkins/prs/version/el_capitan/home/Library/Python/2.7/lib/python/site-packages
  echo 'import site; site.addsitedir("/usr/local/lib/python2.7/site-packages")' >> /Users/brew/Jenkins/prs/version/el_capitan/home/Library/Python/2.7/lib/python/site-packages/homebrew.pth
==> Summary
/usr/local/Cellar/llvm/3.6.2: 1,350 files, 338.2M, built in 18 minutes 27 seconds
```

Unless there's a way to download the full build log...?

**Comment by iMichka** [comment]
I also tried to build castxml on my i5, by passing manually some flags like ```-march=core2```. No way to get a working binary. I think your discoveries for llvm are what we need to explore more.

Castxml needs a llvm/clang install tree (which apple does not provide). Once it is built, the llvm/clang build tree is not needed anymore. You can then use castxml independently with clang (or gcc or mscv).
If ```brew install --build-bottle llvm --with-clang; brew install --build-bottle castxml```worked, this is a good sign.

**Comment by iMichka** [comment]
+1 on the diagnostic.

I did:
```brew install --build-bottle llvm --with-clang```
```brew install --build-bottle castxml```

I then got a ```vxorps```free castxml binary, which ran on my Core2Duo.

**Comment by mikemcquaid** [comment]
> think this means we need to do something about propagating --build-bottle to dependencies which are built from source as part of a bottle build. (Including from-source dependencies pulled in via intermediate bottles in the dependency tree, if such a thing is possible.)

Good find. Probably if we're doing `--build-bottle` then everything should use those CFLAGS (but they'll want to ensure `post_install` is still run).

**Comment by MikeMcQuaid** [comment]
If this is still broken: we'll accept PRs for this but we're not actively working on it at this time.

**Comment by iMichka** [comment]
@MikeMcQuaid I have a workaround on my side for castxml, so I do not need this to be fixed. It's OK for me to close this issue.

Other formulas may still suffer from this. People interested in fixing this may push this forward by themselves.

Also, Core2Duo and Nehalem Xeon's are being slowly phased out so this will probably not be a problem in the future.
